### ISSUE TITLE
Graph rebalances chaotically

### ISSUE THREAD

**Issue by clux** [tchayen/markdown-links#64]
Hey there, thanks for this extension! It's been super useful in helping me visualise a pretty large link network.

I have noticed that between every `Show Graph`, the resulting graph from markdown links is always completely different. This isn't usually a problem, it's interesting to see your data in different ways, but the "default" graph type can often overlay lines in a way that some other edges or vertices are completely obscured. It's also hard to grab screenshots for someone when you have to re-run show-graph, then zoom in on the area you want, and repeat until it displays it in a representative way.

This is probably a hard feature, but do you think that perhaps the default graph algorithm can benefit from some kind of determinism that goes through some rebalancing like in the [d3.js force testing ground](https://bl.ocks.org/steveharoz/8c3e2524079a8c440df60c1ab72b5d03)? Perhaps with exposed force parameters etc.

Do you think this is a possible thing to do statically?

**Comment by clux** [comment]
Ah, I see this is mostly a duplicate of #7, sorry about that.

**Comment by tchayen** [comment]
No problem! Thanks for bringing this up!

So determinism would be one thing. We can go to the source of `d3-force`, replace any `Math.random()` calls and we would get the same graph for the same set of nodes and edges.

The problem lies in how the algorithm works – all points start in the center and their order does not influence where they end up. So You will have a deterministic way of rendering a graph, rendering the same graph with addition of one node, but according to the algorithm those two might end up looking entirely different.

So getting rid of the chaos IMO should consist of two phases:
- Achieving deterministic layout for consecutive runs of the `Show graph` command.
- Achieving something I'd call iterativeness, guaranteeing only slight differences between a graph and the same one with one node added.

**Comment by clux** [comment]
> So You will have a deterministic way of rendering a graph, rendering the same graph with addition of one node, but according to the algorithm those two might end up looking entirely different.

Is that necessarily a problem? It feels like an algorithm that displays the graph in a deterministic way would have to take into the account the weight of clusters to ensure non-chaotic spread and overlaps. If extra nodes added to a cluster causes one cluster to outweigh another, and thus shifting it, it seems pretty reasonable that the heaviest cluster ought to be shifted to the area with the most space available.

Perhaps it's possible to track the number of recursive children on each node? That way we could pre-sort the nodes by weight before starting to graph it. That could maybe provide some limited form of stability?

**Comment by tchayen** [comment]
> Is that necessarily a problem?

If you are viewing the graph while editing your files and add a new one then your graph might end up looking entirely different and you have to look around to find out where are the files you were looking at. So I consider it to be a very important aspect of the issue.

> would have to take into the account the weight of clusters to ensure non-chaotic spread and overlaps

Yes, but for the same reason entire cluster might jump to the opposite side of the graph after adding a file.

> it seems pretty reasonable that the heaviest cluster ought to be shifted to the area with the most space available.

It sounds reasonable but it doesn't really match usage experience expectations. If you are working with the graph view, have it zoomed in to some cluster and then after adding a file the whole cluster ends up in completely different place and you have to unzoom and look for it then I'd consider it a poor experience.

> Perhaps it's possible to track the number of recursive children on each node?

It is possible for sure but without looking into details of how the simulation algorithm works I can't really say if it would make any difference unfortunately.

**Comment by tchayen** [comment]
Closing as not much happened here since the issue was posted.
